### DISCUSSION TITLE
Convert an array of hex values to 1 byte character to use in processing IDE

### DISCUSSION THREAD

**Discussion by MishalJasmine** [micropython/micropython#11470] [Libraries & Drivers]
data_packet_header = [0x0A, 0xFA, 0x0C, 0x00, 0x02] is the array that I need to send to the processing IDE. 

In processing IDE, the data is recieved in the following manner
`inString = blePort.readChar();`
 `pcProcessData(inString);`

I have converted the data_packet_header  into a array of 1byte using:
`struct.pack('>B',data_packet_header[i])` resulting in [b'\n', b'\xfa',  b'\x0c', b'\x00' , b'\x02']

But when printing the input data in Processing IDE,  its not taking b'\n' as 1 char. instead its  receiving b , ' ,\ , n , ' separately.  

Can I get some help on this?

**Comment by jimmo** [comment]
I think you want `bytes(data_packet_header)`

**Reply by MishalJasmine** [reply]
I have tried this. the issue is still there

**Reply by MishalJasmine** [reply]
when I print the `inString` in processing, it looks like this 

    inString:1
    inString:0
    inString:

    inString:

    inString:2
    inString:5
    inString:0
    inString:

    inString:

    inString:1
    inString:2
    inString:

**Reply by jimmo** [reply]
Please provide more context. How does the data get from you to processing? What do you do in MicroPython to send the data?

e.g. you must be doing something like:

```
payload = bytes(data_packet_header)
uart.write(payload)
```

**Reply by MishalJasmine** [reply]
RP2040 is connected to MAX3001 with which it communicates using SPI. The values are read back from MAX and plotted in  processing IDE.

data_packet_header = [0x0A, 0xFA, 0x0C, 0x00, 0x02]

To send the data to the processing, 

    for i in range(0,len(data_packet_header)):
         print(data_packet_header[i])

if I use UART, it does not send data to processing.

**Reply by jimmo** [reply]
> data_packet_header = [0x0A, 0xFA, 0x0C, 0x00, 0x02]

Yes but what comes after that?

**Reply by MishalJasmine** [reply]
After that there is an array of 12 elements which contains the ecg sample and the Bioz sample. After that a footer is sent which contains data_packet_footer = [0x00, 0x0B]

This is repeated every 8msec

**Reply by 2dof** [reply]
@MishalJasmine  I guess that  " ..provide more context " from @jimmo  question you meant (?):  read ECG data  from MAX via RP2040  and send them to the processing ide/computer  by Bluetooth ( on IDE side read with  your blePort  class.)  ?  You send  'payload'  like that ? : [<header><ecg_daata><footer>]  at once or   separately (from context i understand that all is send separately) ? 

if You can please perform some "debug" info, like : print/ show payload (whole frame) before sending (on RPI side), on processing IDE side print/show inString. 
"when I print the inString in processing ..() )"  -> Yo mean ```pcProcessData()```  function or  shortcut for 'processing ide' ?

**Comment by MishalJasmine** [comment]
The MAX30001 is connected to the RP2040 to collect the ECG data. The data collected is then sent to the processing for plotting. The format in which it is sent is as follows:

```python
ecg_sample = ecg.to_bytes(4,'big') 
bioz_sample = bioz.to_bytes(4,'big')
``` 

Header:
```python
data_packet_header = [0x0A, 0xFA, 0x0C, 0x00, 0x02]
``` 

```python
data_packet[0] = ecg_sample[0]
data_packet[1] = ecg_sample[1]
data_packet[2] = ecg_sample[2]
data_packet[3] = ecg_sample[3]
data_packet[4] = bioz_sample[0]
data_packet[5] = bioz_sample[1]
data_packet[6] = bioz_sample[2]
data_packet[7] = bioz_sample[3]
data_packet[8] = 0xFF
data_packet[9] = 0x00
data_packet[10] = 0x00
data_packet[11] = 0x00
``` 
 
Footer:

```python
data_packet_footer = [0x00, 0x0B]
``` 

The data is streamed to the processing in the following manner

```python
for i in range(0,len(data_packet_header)):
       print(data_packet_header[i])
            
for i in range(0,len(data_packet)):
      print(data_packet[i])
        
for i in range(0,len(data_packet_footer)):
      print(data_packet_footer[i])
``` 

It appears to me the data is not getting streamed to the processing in the form of bytes.

**Comment by MishalJasmine** [comment]
I am facing the same issue as this [link](https://forum.pycom.io/topic/3222/object-with-buffer-protocol-required/3). Tried using `bytes` but the issue still exists.

**Comment by ricksorensen** [comment]
Python's `print()` function is largely for printing human readable representations of the arguments.  Here is an example script which runs on both `CPython` and on `micropython` which tries to show what `print` generates and a possible 'raw' print solution.  The expected raw result is 'AB' followed by a carriage return and new line.
```
import sys

d = [0x41, 0x42, 0x0D, 0x0A]  # 'AB\r\n'
db = bytes(d)
print("d:  int array, human readable")
print(d)
print("db: bytes array, human readable")
print(db)

print("db: bytes array, raw")
nw = sys.stdout.buffer.write(db)  # returns number written
print("| done {}\n\n".format(nw))

print("db: bytes array, raw, one at time")
# note on bytes
#   bytes(n)   ... n an integer, returns a null array of length n
#   bytes(str) ... str a string/iterable return an encoded array of len(str)
#  thus there is a need to convert the single/int byte value to an interable
for i in range(len(d)):
    _ = sys.stdout.buffer.write(bytes([d[i]]))  # or bytes(d[i:i+1]) or ..
print("| done {}\n\n".format(nw))
```
Here are the results on my CPython (Ubuntu 22.04) (Note- i made sure to flush the `print(...flush=True)` to keep order.
```
Python 3.10.6
d:  int array, human readable
[65, 66, 13, 10]
db: bytes array, human readable
b'AB\r\n'
db: bytes array, raw
AB
| done 4


db: bytes array, raw, one at time
AB
| done 4
```

This ran on my RP2040, ESP32C3, and Samd51 (WIO Terminal).  My SAMD21 does not support `sys.stdout.buffer`   Here is RP2040 result:

```
MicroPython v1.20.0-24-g867e4dd3d on 2023-04-27; XIAO RP2040 Local with RP2040
Type "help()" for more information.
>>>
Remount local directory ./ at /remote
>>> import chkbytesmp
d:  int array, human readable
[65, 66, 13, 10]
db: bytes array, human readable
b'AB\r\n'
db: bytes array, raw
AB
| done 4


db: bytes array, raw, one at time
AB
| done 4
```

`

**Comment by MishalJasmine** [comment]
yes.. I did observe that the `\r\n` gets appended to every byte that it streamed. If the number sent is `250`, processing IDE is receiving it as `b'250\r\n'`. and it considers `2`, `5`, `0`, `/r` and `/n` as a single byte.
